tools/purge_ems.rb

```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env ruby
+enviroment = File.expand_path('../config/environment', __dir__) + ".rb"
+require enviroment if File.exist?(enviroment)
+
+if !defined?(Vmdb::Application)
+  warn "This script must exist in the tools directory or can live elsewhere but needs to be run via rails runner path/to/purge_ems.rb --id 1234 --follow true"
+  exit 1
+end
+
+require 'optimist'
+
+ARGV.shift if ARGV[0] == '--' # if invoked with rails runner
+opts = Optimist.options do
+  banner "Purge managements system records in the background via the queue.\n\nUsage: ruby #{$0} [options]\n\nOptions:\n\t"
+  opt :id,      "Mangement System id",                                         :short => :i, :type => :integer,                    :required => true
+  opt :follow,  "Follow the progress or exit after creating messages",         :short => :f, :type => :string,  :default => "true"
+  opt :batch,   "How many rows in each relationship to be deleted in a batch", :short => :b, :type => :integer, :default => 1_000
+  opt :timeout, "How many minutes should each message be allowed to run",      :short => :t, :type => :integer, :default => 30
+end
+
+Optimist.die :id,      "must be a positive number"                             if opts[:id] < 1
+Optimist.die :batch,   "must be a positive number"                             if opts[:batch] < 1
+Optimist.die :timeout, "must be a positive number greater than or equal to 10" if opts[:timeout] < 10
+
+opts[:follow] = opts[:follow].to_s.downcase
+Optimist.die :follow, "must be true or false" unless %w[true false].include?(opts[:follow])
+
+RELATIONSHIP_DESTROY_METHOD = "destroy".freeze
+EMS_DESTROY_METHOD          = "destroy_queue".freeze
+ZONE                        = nil                    # Will queue for any zone
+PRIORITY                    = MiqQueue::LOW_PRIORITY # Don't block other work
+
+def log(msg)
+  $log.info("MIQ(#{__FILE__}) #{msg}")
+  puts msg
+end
+
+def current_backlog
+  MiqQueue.where(:method_name => [EMS_DESTROY_METHOD, RELATIONSHIP_DESTROY_METHOD], :zone => ZONE).count
+end
+
+id = opts[:id]
+ems = ExtManagementSystem.find(id)
+if ems.enabled?
+  log("Pausing management system with id: #{id} #{ems.name}...")
+  ems.pause!
```

Comment:
Okay I'm done with bikeshedding, now real questions :laughing: 

We are pausing the EMS here but we are skipping the step in orchestrate_destroy that brings down the workers which is critical.  If there is e.g. a RefreshWorker still running it will likely start bringing back the EMS as unpaused and/or recreate records depending on timing.

I think we need to do the `ems_workers.each(&:kill_async)` and `wait_for_ems_workers_removal` steps here after pausing.

Reply:
> We are pausing the EMS here but we are skipping the step in orchestrate_destroy that brings down the workers which is critical. If there is e.g. a RefreshWorker still running it will likely start bringing back the EMS as unpaused and/or recreate records depending on timing.

Ah, good eye.  We're not actually "skipping" the step.  We're doing `Ems#destroy_queue` which tosses another `pause!` on the queue and a `orchestrate_destroy_queue`.  Unfortunately, the latter does both the part you're concerned about and the `Ems#destroy` in line so I guess we want to get in between that, so we can destroy the relationships after the workers are gone.

We probably want to take parts of this script and update the code to split the destroy from this pre-destroy work.

For now, we can pull in these two steps.

So, sync:

```
ems.pause!
ems.ems_workers.each(&:kill_async)
ems.wait_for_ems_workers_removal
...
# queue up the ems relationship destroy in batches
```

Reply:
> so I guess we want to get in between that, so we can destroy the relationships after the workers are gone.

Yeah exactly I think we want to make sure the workers are down before we start destroying the relationships since the `Ems#destroy_queue` is queued after all the realt destroys

Reply:
Ok, pushed changes... let me know if it looks better

Reply:
:+1: looks good